Add documentation for "Running Containers"

ajeetraina opened the pull request:
This PR introduces a new collection of guides aimed at establishing a strong foundation in "Running containers". This initial set of guides focuses on:

- Publishing ports
- Overriding container defaults
- Persisting container data
- Multi-container applications

craig-osterhout commented on content/guides/docker-concepts/running-containers/overriding-container-defaults.md:
```diff
@@ -0,0 +1,177 @@
+
+In this hands-on guide, you'll see how to use the `docker run` command to override the container defaults.
+
+1. [Download and install](https://www.docker.com/products/docker-desktop/) Docker Desktop.
```
Either remove the numbered list here, or add this as the first step in the section below.
```suggestion
1. [Download and install](/get-docker/) Docker Desktop.
```

craig-osterhout commented on content/guides/docker-concepts/running-containers/overriding-container-defaults.md:
```diff
@@ -0,0 +1,177 @@
+
+    This will start a new Redis container in the background, listening on the standard container port `6379` but mapped to port `6380` on the host machine. You override the host port just to ensure that this new container doesn't conflict with the existing running container.
+
+3. Verify if both containers are running via the Docker dashboard
```
```suggestion
3. Verify if both containers are running via the Docker Dashboard.
```

craig-osterhout commented on content/guides/docker-concepts/running-containers/overriding-container-defaults.md:
```diff
@@ -0,0 +1,177 @@
+
+### Manage the resources
+
+By default, containers are not limited in their resource usage. However, on shared systems, it's crucial to manage resources effectively. It's important not to allow a running container to consume too much of the host machine's memory.
```
```suggestion
By default, containers are not limited in their resource usage. However, on shared systems, it's crucial to manage resources effectively. It's important not to let a running container consume too much of the host machine's memory.
```

craig-osterhout commented on content/guides/docker-concepts/running-containers/overriding-container-defaults.md:
````diff
@@ -0,0 +1,177 @@
+$ docker run -d --memory=”512m” --cpus=”.5” redis
+```
+
+The `--cpus` flag specifies the CPU quota for the container. Here, it's set to half a CPU core (0.5) whereas the `--memory` flag specifies the memory limit for the container. In this case, it's set to 512 megabytes (512MB).
````
```suggestion
The `--cpus` flag specifies the CPU quota for the container. Here, it's set to half a CPU core (0.5) whereas the `--memory` flag specifies the memory limit for the container. In this case, it's set to 512 MB.
```

craig-osterhout commented on content/guides/docker-concepts/running-containers/overriding-container-defaults.md:
```diff
@@ -0,0 +1,177 @@
+
+### Override the default `CMD` and `ENTRYPOINT` in Docker Compose
+
+Sometimes, you might need to override the default commands (CMD) or entry points (ENTRYPOINT) defined in a Docker image, especially when using Docker Compose.
```
```suggestion
Sometimes, you might need to override the default commands (`CMD`) or entry points (`ENTRYPOINT`) defined in a Docker image, especially when using Docker Compose.
```

craig-osterhout commented on content/guides/docker-concepts/running-containers/overriding-container-defaults.md:
```diff
@@ -0,0 +1,177 @@
+
+Sometimes, you might need to override the default commands (CMD) or entry points (ENTRYPOINT) defined in a Docker image, especially when using Docker Compose.
+
+1. Create a `docker-compose.yaml` file with the following content:
```
```suggestion
1. Create a `compose.yaml` file with the following content:
```

craig-osterhout commented on content/guides/docker-concepts/running-containers/overriding-container-defaults.md:
````diff
@@ -0,0 +1,177 @@
+         command: ["redis-server", "--requirepass", "redispassword"]
+    ```
+
+    The YAML file defines a service named `redis` that uses the official Redis image, sets an entrypoint script, and starts the container with password authentication.
````
```suggestion
    The Compose file defines a service named `redis` that uses the official Redis image, sets an entrypoint script, and starts the container with password authentication.
```

craig-osterhout commented on content/guides/docker-concepts/running-containers/overriding-container-defaults.md:
```diff
@@ -0,0 +1,177 @@
+
+    This command starts the Redis service defined in the Docker Compose file.
+
+3. Verify the authentication with Docker Dashboard
```
```suggestion
3. Verify the authentication with Docker Dashboard.
```

craig-osterhout commented on content/guides/docker-concepts/running-containers/overriding-container-defaults.md:
```diff
@@ -0,0 +1,177 @@
+
+3. Verify the authentication with Docker Dashboard
+
+    Open the **Docker Dashboard**, select the **Redis** container and select **Exec** to enter into the container shell.
```
```suggestion
    Open the Docker Dashboard, select the **Redis** container and select **Exec** to enter into the container shell.
```

craig-osterhout commented on content/guides/docker-concepts/running-containers/overriding-container-defaults.md:
```diff
@@ -0,0 +1,177 @@
+
+## Additional resources
+
+* [Build Variables](https://docs.docker.com/build/building/variables/)
```
```suggestion
* [Build variables](/build/building/variables/)
```

craig-osterhout commented on content/guides/docker-concepts/running-containers/overriding-container-defaults.md:
```diff
@@ -0,0 +1,177 @@
+## Additional resources
+
+* [Build Variables](https://docs.docker.com/build/building/variables/)
+* [Ways to set environment variables with Compose](https://docs.docker.com/compose/environment-variables/set-environment-variables/)
```
```suggestion
* [Ways to set environment variables with Compose](/compose/environment-variables/set-environment-variables/)
```

craig-osterhout commented on content/guides/docker-concepts/running-containers/persisting-container-data.md:
```diff
@@ -0,0 +1,176 @@
+---
+title: Persisting container data
+keywords: concepts, build, images, container, docker desktop
+description: This concept page will teach you the significance of data persistent in Docker
```
```suggestion
description: This concept page will teach you the significance of data persistence in Docker
```

craig-osterhout commented on content/guides/docker-concepts/running-containers/persisting-container-data.md:
````diff
@@ -0,0 +1,176 @@
+> While you can use the `docker volume create` command to create a volume, Docker will automatically create a volume when needed. For example, the following command will auto-create a volume named `sample-volume` if it doesn’t exist:
+>
+> ```console
+> $ docker run -v sample-volumes:/data my-image
````
```suggestion
> $ docker run -v sample-volume:/data my-image
```

craig-osterhout commented on content/guides/docker-concepts/running-containers/persisting-container-data.md:
```diff
@@ -0,0 +1,176 @@
+
+### Use volumes
+
+1. [Download and instal](https://www.docker.com/products/docker-desktop/)l Docker Desktop.
```
```suggestion
1. [Download and install](/get-docker/) Docker Desktop.
```

craig-osterhout commented on content/guides/docker-concepts/running-containers/persisting-container-data.md:
````diff
@@ -0,0 +1,176 @@
+
+4. In the PostgreSQL command line, run the following to create a database table and insert two records:
+
+    ```psql
````
The formatting doesn't look great. Maybe consider using `text` instead of `psql`.

ajeetraina replied:
I changed it to "text".

ajeetraina replied:
Fixed.

craig-osterhout commented on content/guides/docker-concepts/running-containers/persisting-container-data.md:
````diff
@@ -0,0 +1,176 @@
+
+5. Verify the data is in the database by running the following in the PostgreSQL command line:
+
+    ```psql
````
The formatting doesn't look great. Maybe consider using `text` instead of `psql`.

ajeetraina replied:
Fixed it. Can you check it once?

craig-osterhout commented on content/guides/docker-concepts/running-containers/persisting-container-data.md:
```diff
@@ -0,0 +1,176 @@
+
+The Docker Dashboard provides the ability to view the contents of any volume, as well as the ability to export, import, and clone volumes.
+
+1. Open the **Docker Dashboard** and navigate to the **Volumes** view. In this view, you should see the **postgres_data** volume
```
```suggestion
1. Open the Docker Dashboard and navigate to the **Volumes** view. In this view, you should see the **postgres_data** volume.
```

craig-osterhout commented on content/guides/docker-concepts/running-containers/persisting-container-data.md:
```diff
@@ -0,0 +1,176 @@
+
+2. Select the **postgres_data** volume’s name.
+
+3. The **Data** tab shows the contents of the volume and provides the ability to navigate the files. Double-clicking on a file will allow you to see the contents and make changes.
```
```suggestion
3. The **Data** tab shows the contents of the volume and provides the ability to navigate the files. Double-clicking on a file will let you see the contents and make changes.
```

craig-osterhout commented on content/guides/docker-concepts/running-containers/persisting-container-data.md:
```diff
@@ -0,0 +1,176 @@
+
+There are a few methods to remove volumes, including the following:
+
+1. Select the **Delete Volume** option on a volume in the **Docker Dashboard**.
```
```suggestion
1. Select the **Delete Volume** option on a volume in the Docker Dashboard.
```

craig-osterhout commented on content/guides/docker-concepts/running-containers/persisting-container-data.md:
```diff
@@ -0,0 +1,176 @@
+
+1. Select the **Delete Volume** option on a volume in the **Docker Dashboard**.
+
+2. Use the `docker volume rm` command
```
```suggestion
2. Use the `docker volume rm` command:
```

craig-osterhout commented on content/guides/docker-concepts/running-containers/persisting-container-data.md:
````diff
@@ -0,0 +1,176 @@
+    $ docker volume rm postgres_data
+    ```
+
+3. Use the `docker volume prune` command to remove all unused volumes
````
```suggestion
3. Use the `docker volume prune` command to remove all unused volumes:
```

craig-osterhout commented on content/guides/docker-concepts/running-containers/persisting-container-data.md:
```diff
@@ -0,0 +1,176 @@
+
+There are a few methods to remove volumes, including the following:
+
+1. Select the **Delete Volume** option on a volume in the **Docker Dashboard**.
```
Typically ordered lists are for things that must be done in sequence. Consider using a bulleted list for options where only one thing needs to be done.

ajeetraina replied:
Thanks for the suggestion. I have added the bullet points instead.

craig-osterhout commented on content/guides/docker-concepts/running-containers/persisting-container-data.md:
```diff
@@ -0,0 +1,176 @@
+
+- [Manage data in Docker](/storage)
+- [Volumes](/storage/volumes)
+- [Volume mounts (docker run reference)](/engine/reference/run/#volume-mounts)
```
```suggestion
- [Volume mounts (`docker run` reference)](/engine/reference/run/#volume-mounts)
```

craig-osterhout commented on content/guides/docker-concepts/running-containers/multi-container-applications.md:
```diff
@@ -0,0 +1,203 @@
+- Imagine running several `docker run` commands (frontend, backend, and database) with different configurations for development, testing, and production environments. It's error-prone and time-consuming.
+- Applications often rely on each other. Manually starting containers in a specific order and managing network connections become difficult as the stack expands.
+- Each application needs its `docker run` command, making it difficult to scale individual services. Scaling the entire application means potentially wasting resources on components that don't need a boost.
+- Persisting data for each application requires separate volume mounts or configurations within each docker run command. This creates a scattered data management approach.
```
```suggestion
- Persisting data for each application requires separate volume mounts or configurations within each `docker run` command. This creates a scattered data management approach.
```

craig-osterhout commented on content/guides/docker-concepts/running-containers/multi-container-applications.md:
```diff
@@ -0,0 +1,203 @@
+- Applications often rely on each other. Manually starting containers in a specific order and managing network connections become difficult as the stack expands.
+- Each application needs its `docker run` command, making it difficult to scale individual services. Scaling the entire application means potentially wasting resources on components that don't need a boost.
+- Persisting data for each application requires separate volume mounts or configurations within each docker run command. This creates a scattered data management approach.
+- Setting environment variables for each application through separate docker run commands is tedious and error-prone.
```
```suggestion
- Setting environment variables for each application through separate `docker run` commands is tedious and error-prone.
```

craig-osterhout commented on content/guides/docker-concepts/running-containers/multi-container-applications.md:
```diff
@@ -0,0 +1,203 @@
+
+Docker Compose defines your entire multi-container application in a single YAML file called `compose.yml`. This file specifies configurations for all your containers, their dependencies, environment variables, and even volumes and networks. With Docker Compose:
+
+- You don't need to run multiple docker run commands. All you need to do is define your entire multi-container application in a single YAML file. This centralizes configuration and simplifies management.
```
```suggestion
- You don't need to run multiple `docker run` commands. All you need to do is define your entire multi-container application in a single YAML file. This centralizes configuration and simplifies management.
```

craig-osterhout commented on content/guides/docker-concepts/running-containers/multi-container-applications.md:
```diff
@@ -0,0 +1,203 @@
+
+## Try it now
+
+In this hands-on guide, you'll first see how to build and run a counter web application based on Node.js, Nginx reverse proxy and Redis database using the `docker run` commands. You’ll also see how you can simplify the entire deployment process using Docker Compose.
```
```suggestion
In this hands-on guide, you'll first see how to build and run a counter web application based on Node.js, an Nginx reverse proxy, and a Redis database using the `docker run` commands. You’ll also see how you can simplify the entire deployment process using Docker Compose.
```

craig-osterhout commented on content/guides/docker-concepts/running-containers/multi-container-applications.md:
```diff
@@ -0,0 +1,203 @@
+   {{< /tab >}}
+   {{< /tabs >}}
+
+2. Navigate into the `nginx-node-redis-main` directory
```
```suggestion
2. Navigate into the `nginx-node-redis-main` directory:
```

craig-osterhout commented on content/guides/docker-concepts/running-containers/multi-container-applications.md:
```diff
@@ -0,0 +1,203 @@
+
+    Inside this directory, you'll find two sub-directories - `nginx` and `web`.
+
+3. [Download and install](https://www.docker.com/products/docker-desktop/) Docker Desktop.
```
```suggestion
3. [Download and install](/get-docker/) Docker Desktop.
```

craig-osterhout commented on content/guides/docker-concepts/running-containers/multi-container-applications.md:
````diff
@@ -0,0 +1,203 @@
+    $ docker run -d --name nginx --network sample-app  -p 80:80 nginx
+    ```
+
+     > Nginx is typically used as a reverse proxy for web applications, routing traffic to backend servers. In our case, it routes to the Node.js backend containers (web1 or web2)
````
Suggest using a typical note with a title.

```suggestion
     > **Note**
     >
     > Nginx is typically used as a reverse proxy for web applications, routing traffic to backend servers. In this case, it routes to the Node.js backend containers (web1 or web2).
```

craig-osterhout commented on content/guides/docker-concepts/running-containers/multi-container-applications.md:
````diff
@@ -0,0 +1,203 @@
+
+6.  Verify the containers are up by running the following command:
+
+    ```console
````
Suggest separating input and output into different codeblocks because it makes it easier to use the copy button.

ajeetraina replied:
Fixed.

craig-osterhout commented on content/guides/docker-concepts/running-containers/multi-container-applications.md:
````diff
@@ -0,0 +1,203 @@
+    web1: Number of visits is: 12
+    ```
+
+    > You might have noticed that Nginx, acting as a reverse proxy, likely distributes incoming requests in a round-robin fashion between the two backend containers. This means each request might be directed to a different container (web1 and web2) on a rotating basis. The output shows consecutive increments for both the web1 and web2 containers and the actual counter value stored in Redis is upd ...
````
Suggest adding a title.
```suggestion
    > **Note**
    >
    > You might have noticed that Nginx, acting as a reverse proxy, likely distributes incoming requests in a round-robin fashion between the two backend containers. This means each request might be directed to a different container (web1 and web2) on a rotating basis. The output shows consecutive increments for both the web1 and web2 containers and the actual counter value stored in Redis is updated only after the response is sent back to the client.
```

craig-osterhout commented on content/guides/docker-concepts/running-containers/multi-container-applications.md:
```diff
@@ -0,0 +1,203 @@
+
+    > You might have noticed that Nginx, acting as a reverse proxy, likely distributes incoming requests in a round-robin fashion between the two backend containers. This means each request might be directed to a different container (web1 and web2) on a rotating basis. The output shows consecutive increments for both the web1 and web2 containers and the actual counter value stored in Redis is upd ...
+
+9. You can use the Docker Dashboard to remove the containers by selecting the containers and selecting the *Delete* button
```
```suggestion
9. You can use the Docker Dashboard to remove the containers by selecting the containers and selecting the **Delete** button
```

craig-osterhout commented on content/guides/docker-concepts/running-containers/multi-container-applications.md:
```diff
@@ -0,0 +1,203 @@
+Docker Compose provides a structured and streamlined approach for managing multi-container deployments. As stated earlier, with Docker Compose, you don’t need to run multiple `docker run` commands. All you need to do is define your entire multi-container application in a single YAML file called `compose.yml`. Let’s see how it works.
+
+
+Navigate to the root of the project directory. Inside this directory, you'll find a file named `compose.yml`. This YAML file is where all the magic happens! It defines all the services that make up your application, along with their configurations. Each service specifies its image, ports, volumes, networks, and any other settings necessary for its functionality.
```
```suggestion
Navigate to the root of the project directory. Inside this directory, you'll find a file named `compose.yml`. This YAML file is where all the magic happens. It defines all the services that make up your application, along with their configurations. Each service specifies its image, ports, volumes, networks, and any other settings necessary for its functionality.
```

craig-osterhout commented on content/guides/docker-concepts/running-containers/multi-container-applications.md:
```diff
@@ -0,0 +1,203 @@
+9. You can use the Docker Dashboard to remove the containers by selecting the containers and selecting the *Delete* button
+
+
+    ![A screenshot of Docker Dashboard showing multi-container applications](images/multi-container-apps.webp?border=true)
```
Verify that this is the correct image.

ajeetraina replied:
Added the right images.

craig-osterhout commented on content/guides/docker-concepts/running-containers/multi-container-applications.md:
```diff
@@ -0,0 +1,203 @@
+2. If you look at the Docker Dashboard, you can see the containers and dive deeper into their configuration.
+
+
+    ![A screenshot of Docker Dashboard showing how to delete the multi-container applications](images/delete-multi-container-apps.webp?border=true)
```
Verify that this is the correct image.

ajeetraina replied:
A Good catch. Added the right images at the right location.

craig-osterhout commented on content/guides/docker-concepts/running-containers/multi-container-applications.md:
```diff
@@ -0,0 +1,203 @@
+
+
+
+3. Alternatively, you can use the Docker Dashboard to remove the containers by selecting the application stack and selecting the Delete button.
```
```suggestion
3. Alternatively, you can use the Docker Dashboard to remove the containers by selecting the application stack and selecting the **Delete** button.
```

craig-osterhout commented on content/guides/docker-concepts/running-containers/multi-container-applications.md:
```diff
@@ -0,0 +1,203 @@
+
+## Additional resources
+
+* [Docker container run](https://docs.docker.com/reference/cli/docker/container/run/)
```
```suggestion
* [`docker container run` CLI reference](reference/cli/docker/container/run/)
```

craig-osterhout commented on content/guides/docker-concepts/running-containers/multi-container-applications.md:
```diff
@@ -0,0 +1,203 @@
+## Additional resources
+
+* [Docker container run](https://docs.docker.com/reference/cli/docker/container/run/)
+* [Run multi-container applications](https://docs.docker.com/guides/walkthroughs/multi-container-apps/)
```
```suggestion
* [Run multi-container applications](/guides/walkthroughs/multi-container-apps/)
```

ajeetraina commented:
> Looking good. Added some suggestions. Hopefully the vale test will pass when the items here are fixed. Also, there are some rogue files (`the-basics/what-is-co` and `images/over`) in this PR that should be deleted.

I have removed these rogue files.

ajeetraina commented:
@craig-osterhout As you suggested, the vale test now shows PASS.

craig-osterhout commented on content/guides/docker-concepts/running-containers/overriding-container-defaults.md:
````diff
@@ -0,0 +1,180 @@
+3. Connect Redis to the existing network by using the following command:
+
+    ```console
+    $ docker run -d -p 6381:6379 –network mynetwork redis
````
```suggestion
    $ docker run -d -p 6381:6379 --network mynetwork redis
```

craig-osterhout commented on content/guides/docker-concepts/running-containers/overriding-container-defaults.md:
````diff
@@ -0,0 +1,180 @@
+This is where the `docker run` command shines again. It offers flags like `--memory` and `--cpus` to restrict how much CPU and memory a container can use.
+
+```console
+$ docker run -d --memory=”512m” --cpus=”.5” redis
````
```suggestion
$ docker run -d --memory="512m" --cpus=".5" redis
```

craig-osterhout commented on content/guides/docker-concepts/running-containers/persisting-container-data.md:
````diff
@@ -0,0 +1,174 @@
+9. Verify the database still has the records by running the following command:
+
+    ```console
+    $ docker exec -ti new-db psql -U postgres -c “SELECT * FROM tasks”
````
```suggestion
    $ docker exec -ti new-db psql -U postgres -c "SELECT * FROM tasks"
```

craig-osterhout commented on content/guides/docker-concepts/running-containers/multi-container-applications.md:
```diff
@@ -0,0 +1,210 @@
+    >
+    > You might have noticed that Nginx, acting as a reverse proxy, likely distributes incoming requests in a round-robin fashion between the two backend containers. This means each request might be directed to a different container (web1 and web2) on a rotating basis. The output shows consecutive increments for both the web1 and web2 containers and the actual counter value stored in Redis is upd ...
+
+9. You can use the Docker Dashboard to remove the containers by selecting the containers and selecting the **Delete** button
```
```suggestion
9. You can use the Docker Dashboard to remove the containers by selecting the containers and selecting the **Delete** button.
```

craig-osterhout reviewed (changes requested):
Thanks Ajeet. I missed a few small things in the initial review. The smart quotes and one of the dashes in the codeblocks didn't copy-paste good. Other than that LGTM.

ajeetraina commented:
Thanks @craig-osterhout I went through all the reviewed comments.

dvdksn commented on content/guides/docker-concepts/running-containers/overriding-container-defaults.md:
````diff
@@ -0,0 +1,180 @@
+Sometimes you might want to use separate database instances for development and testing purposes. Running these database instances on the same port might conflict. You can use the `-p` option in `docker run` to map container ports to host ports, allowing you to run the multiple instances of the container without any conflict.
+
+```console
+$ docker run -d -p HOST_PORT:CONTAINER_PORT my_image
````
I don't think we should use `my_image` here and in the following examples. Let's use `alpine` instead (and add a brief note about it in the introduction)

ajeetraina replied:
As the initial section talks about the database, I picked up postgres Docker image instead rather than my_image. Do check if it looks good. @dvdksn

ajeetraina replied:
@dvdksn I have changed all the my_image to the real world Docker images. Can you please have a look?

dvdksn commented on content/guides/docker-concepts/running-containers/overriding-container-defaults.md:
```diff
@@ -0,0 +1,180 @@
+
+### Run the multiple instance of Redis database
+
+1.  Start a container using the [Redis image](https://hub.docker.com/_/redis) with the following command:
```
Given the recent licensing events about redis, let's maybe use a different example image. I propose `memcached`

ajeetraina replied:
@dvdksn I need to test it with Memcached. Will test it tomorrow and let you know.

ajeetraina replied:
I have pushed the revised guide replacing redis with postgres.

ajeetraina replied:
Also, I am checking internally to see if we can replace the video with Postgres example.

dvdksn commented on content/guides/docker-concepts/running-containers/overriding-container-defaults.md:
```diff
@@ -0,0 +1,180 @@
+
+## Additional resources
+
+* [Build variables](/build/building/variables/)
```
This page doesn't have anything related to build variables so I don't think we need this link. [This page](https://docs.docker.com/engine/reference/run/) has a lot of overlap with this guide so maybe somewhere there is where you want to link to.

ajeetraina replied:
I have removed the link.

ajeetraina commented on content/guides/docker-concepts/running-containers/images/delete-the-container.webp:
Removed.

ajeetraina commented on content/guides/docker-concepts/running-containers/images/deleted-files.webp:
Removed this unused image.

ajeetraina commented on content/guides/docker-concepts/running-containers/images/exec-into-redis-container.webp:
Removed this unused image.

ajeetraina commented on content/guides/docker-concepts/running-containers/images/mounted-files.webp:
Removed this unused image.

ajeetraina commented on content/guides/docker-concepts/running-containers/images/running-redis-containers.webp:
Removed this unused image.

ajeetraina commented on content/guides/docker-concepts/running-containers/images/access-the-frontend.webp:
Done. Pointed to the existing image

ajeetraina commented on content/guides/docker-concepts/running-containers/overriding-container-defaults.md:
````diff
@@ -0,0 +1,187 @@
+This option sets an environment variable `foo` inside the container with the value `bar`.
+
+```console
+$ docker run -e foo=bar postgres env
````
If you look at the initial part, it says "For example, if you have an existing database container that listens on the standard port and you want to run a new instance of the same database container, then you might want to change the port settings the new container listens on so that it doesn’t conflict with the existing container." As we're talking about a database container, it does make sense to pick up a database container rather than Alpine. Changing to alpine would disturb the flow of the explanation. Let me know your thoughts.

ajeetraina commented on content/guides/docker-concepts/running-containers/overriding-container-defaults.md:
```diff
@@ -0,0 +1,187 @@
+> **Tip**
+>
+> The `.env` file acts as a convenient way to set environment variables for your Docker containers without cluttering your command line with numerous `-e` flags. To use a `.env` file, you can pass `--env-file` option with the `docker run` command.
+{ .tip }
```
Added to the tip.

ajeetraina commented on content/guides/docker-concepts/running-containers/persisting-container-data.md:
````diff
@@ -0,0 +1,174 @@
+>
+> ```console
+> $ docker run -d -p 80:80 -v sample-volume:/data docker/welcome-to-docker
+> ```
````
I have removed the call out and put it in a simple paragraph.

ajeetraina commented on content/guides/docker-concepts/running-containers/images/list-containers.webp:
Uploaded the better image.

ajeetraina commented on content/guides/docker-concepts/running-containers/images/multi-container-apps.webp:
Uploaded the better image.

ajeetraina commented on content/guides/docker-concepts/running-containers/images/delete-containers.webp:
Uploaded the better image.

ajeetraina commented on content/guides/docker-concepts/running-containers/images/published-ports.webp:
Uploaded the better image.

ajeetraina commented:
@dvdksn I went through all your comments and added my review. For multi-container apps, I will be uploading a new video based on Postgres.

dvdksn commented:
Thank you @ajeetraina 

Looks like there are a couple lint errors too

```
8.071 content/guides/docker-concepts/running-containers/multi-container-applications.md:86:1 MD029/ol-prefix Ordered list item prefix [Expected: 2; Actual: 3; Style: 1/2/3]
8.072 content/guides/docker-concepts/running-containers/overriding-container-defaults.md:170 MD040/fenced-code-language Fenced code blocks should have a language specified [Context: "    ```"]
```

ajeetraina commented:
> ntext: "    ```"]

I don't see such error while I use vale.

```
vale overriding-container-defaults.md

 overriding-container-defaults.md
 13:1     warning     Write short, concise            Docker.SentenceLength
                      sentences. (<=30 words)
 13:292   warning     Write short, concise            Docker.SentenceLength
                      sentences. (<=30 words)
 13:535   suggestion  Consider removing 'properly'.   Docker.Adverbs
 44:142   suggestion  Consider using 'many' instead   Docker.ComplexWords
                      of 'numerous'.
 62:167   suggestion  Consider using 'enough'         Docker.ComplexWords
                      instead of 'sufficient'.
 62:186   suggestion  Consider using 'change'         Docker.ComplexWords
                      instead of 'adjustment'.
 125:76   suggestion  'is tied' looks like passive    Docker.Passive
                      voice.
 130:24   suggestion  Consider using 'aren't'         Docker.Contractions
                      instead of 'are not'.
 175:111  warning     'EXEC' has no definition.       Docker.Acronyms
 180:194  suggestion  'be required' looks like        Docker.Passive
                      passive voice.

✖ 0 errors, 3 warnings and 7 suggestions in 1 file.
```

ajeetraina commented:
@dvdksn I couldn't figure out "8.071 content/guides/docker-concepts/running-containers/multi-container-applications.md:86:1 MD029/ol-prefix Ordered list item prefix [Expected: 2; Actual: 3; Style: 1/2/3]" error. I am not sure where it is coming from and what's the fix.

dvdksn commented on content/guides/docker-concepts/running-containers/multi-container-applications.md:
```diff
@@ -0,0 +1,221 @@
+   {{< /tabs >}}
+
+
+3. [Download and install](/get-docker/) Docker Desktop.
```
```suggestion
2. [Download and install](/get-docker/) Docker Desktop.
```

dvdksn commented on content/guides/docker-concepts/running-containers/overriding-container-defaults.md:
````diff
@@ -0,0 +1,194 @@
+
+    Open the Docker Dashboard, select the **Postgres** container and select **Exec** to enter into the container shell. You can type the following command to connect to the Postgres database:
+
+    ```
+    psql -U postgres
````
```suggestion
    ```console
    # psql -U postgres
```

ajeetraina commented on content/guides/docker-concepts/running-containers/images/delete-multi-container-apps.webp:
Fixed.

dvdksn commented on content/guides/docker-concepts/running-containers/overriding-container-defaults.md:
```diff
@@ -0,0 +1,194 @@
+
+    ![A screenshot of Docker Dashboard showing the running instances of Postgres containers](images/running-postgres-containers.webp?border=true)
+
+### Run Postgres container in a controlled network
```
This section doesn't address the difference between the default bridge and custom bridge networks and DNS resolution. The paragraph seems to suggest that the default bridge is convenient for container intercommunication, but that's only via container ip which isn't very usable. I think this section should demonstrate DNS resolution over a custom bridge instead (not only how to connect to one)

ajeetraina replied:
Added a tip/section that discusses customer Vs default bridge network.

dvdksn replied:
Thanks!

dvdksn commented on content/guides/docker-concepts/running-containers/overriding-container-defaults.md:
```diff
@@ -0,0 +1,194 @@
+
+The `--cpus` flag specifies the CPU quota for the container. Here, it's set to half a CPU core (0.5) whereas the `--memory` flag specifies the memory limit for the container. In this case, it's set to 512 MB.
+
+### Override the default `CMD` and `ENTRYPOINT` in Docker Compose
```
I think this section should not only show the compose example, but also overriding default commands for `docker run`.

ajeetraina replied:
I added a section for overriding the container default with`docker run`. Do check and let me know if it looks good.

dvdksn reviewed (dismissed):
LGTM

ajeetraina commented:
@dvdksn Just to let you know that I will be uploading the newer YT video for our overriding container defaults page. Will be raising a PR once I get the final video uploaded.

mikesir87 commented:
@ajeetraina @dvdksn @craig-osterhout  I just uploaded the new YT video, updated the embed link, and verified it's correct in the preview site. Should be good to merge at this point. Thanks all!
